attestedtls/listener.go

```diff
@@ -0,0 +1,159 @@
+package attestedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"time"
+	"github.com/golang/protobuf/proto"
+	// local modules
+	log "github.com/sirupsen/logrus"
+	ci "github.com/Fraunhofer-AISEC/cmc/cmcinterface"
+	// debug
+	"encoding/hex"
+)
+
+var id = "0000"
+var timeout = 10 * time.Second
+
+/* Checks Attestation report by calling the CMC to Verify and checking its status response
+ */
+ func obtainAR(req *ci.AttestationRequest) (resp *ci.AttestationResponse, err error) {
+	// Get backend connection
+	cmcClient, cmcconn, cancel := getCMCServiceConn();
+	if cmcClient == nil {
+		return nil, errors.New("[Listener] Connection failed. No result obtained")
+	}
+	defer cmcconn.Close()
+	defer cancel()
+
+	// Extend Attest request with id
+	req.Id = id
+
+	// Call Attest request
+	resp, err = cmcClient.Attest(context.Background(), req)
+	if err != nil {
+		log.Error(err)
+		return nil, errors.New("[Listener] Could not obtain attestation report")
+	}
+
+	// Return response
+	return resp, nil
+}
+
+/***********************************************************
+* net.Listener Wrapper -> attestedtls.Listener
+*/
+
+/* Struct to implement Listener interface
+ * holds net.Listener and adds additional functionality to its functions */
+type Listener struct {
```

Comment:
Personally, I would make this type unexported and then only return a NewListener using a `net.Listener` interface. This is similar to how the tls package does this internally

See https://github.com/golang/go/blob/2580d0e08d5e9f979b943758d3c49877fb2324cb/src/crypto/tls/tls.go#L58-L82

Reply:
Additionally, similar to the `tls` package, I would use a struct/interface embedding, i.e. leave out the `Ln` name. This way, you do not need to implement the  functions that you are not changing, e.g. `Addr` and `Close`, but those will automatically be taken from the embedded `net.Listener`.

Reply:
Thank you, I will look into that